Feat/2504 add an argillatraining module to the core library with training workflows

davidberenstein1957 opened the pull request:
# Description

Hi, this PR contains some preliminary work on the creating of an `Argilla.training` module, which should simplify training workflows for training using Argilla datasets for several common tasks using several common libraries. Initially, it only contains a `ArgillaSetFitTrainer` and an `ArgillaBaseTrainer`.

Closes #2504 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

Not yet implemented tests

**Checklist**

- [ ] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

davidberenstein1957 committed:
chore: initial outline for the argilla.training module #2504

davidberenstein1957 committed:
chore: first excecutable version

davidberenstein1957 committed:
chore: updated structure w.r.t. passing *args and **kwargs

davidberenstein1957 committed:
chore: cleaned code and added `seed` param to trainer

davidberenstein1957 committed:
chore: added docstrings to argilla.training.utils

davidberenstein1957 committed:
chore: added multi-label support and minor prepare_for_training bugfixes

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+from argilla.training.setfit import ArgillaSetFitTrainer
+
+
+class ArgillaBaseTrainer(object):
```
```suggestion
class ArgillaTrainer:
```
No need to specify that it is a Base trainer. If there's no qualifier (`ArgillaSetFitTrainer`, `ArgillaSpaCyTrainer`, `ArgillaTransformersTrainer`), then we already know that it's the "base" one.
Also, for Python 3 we don't need to subclass `object` anymore.

davidberenstein1957 replied:
Thanks, resolved this.

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+                self.device = "cpu"
+
+        if framework == "setfit":
+            assert self._rg_dataset_type == rg.DatasetForTextClassification, "SetFit supports only text classification"
```
Generally it's not recommended to use assert's in actual code, as you can use compiler flags to remove asserts. We should probably just use an `if` with a `raise`.

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+            elif torch.cuda.is_available():
+                self.device = "cuda"
+            else:
+                self.device = "cpu"
```
Perhaps we can defer this logic to the more specific subclasses (`ArgillaSetFitTrainer`, `ArgillaTransformersTrainer`, etc.) instead? That way the base trainer needs to know less about the subclasses.

davidberenstein1957 replied:
I wanted to avoid copying this logic at least twice  (transformers, setfit). Maybe also spacy[transformers]. But I agree and moved it to the SetFit custom logic for now.

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+            raise NotImplementedError(f"Framework {framework} is not supported")
+        self._logger.warning(self)
+
+    def __repr__(self) -> str:
```
I love this, well done.

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+    def predict(self, text: Union[List[str], str], as_argilla_records: bool = True):
+        """
+        `predict` takes a string or list of strings and returns a list of dictionaries, each dictionary
+        containing the text, the predicted label, and the confidence score
```
```suggestion
        containing the text, the predicted label, and the confidence score.
```

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+        Args:
+          text (Union[List[str], str]): The text to be classified.
+          as_argilla_records (bool): If True, the output will be a list of ArgillaRecord objects. If
+        True, the output will be a list of Argilla records. Defaults to True.
```
```suggestion
          as_argilla_records (bool): If True, the output will be a list of Argilla records instead of dictionaries. Defaults to True.
```
Duplicated, and my update mentions the dictionaries, too.

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+        True, the output will be a list of Argilla records. Defaults to True.
+
+        Returns:
+          A list of predictions or argilla records.
```
```suggestion
          A list of predictions or Argilla records.
```

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+    def train(self, path: str = None):
+        """
+        > The function `train` takes in a path to a file and trains the model. If a path is provided,
+        the model is saved to that path
```
```suggestion
        `train` takes in a path to a file and trains the model. If a path is provided,
        the model is saved to that path.
```

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+        """
+        self._trainer.train()
+        if path is not None:
+            self._trainer.save(path)
```
```suggestion
            self.save(path)
```
This might be preferable?

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,163 @@
+
+    def save(self, path: str):
+        """
+        It saves the model to the path specified
```
```suggestion
        Saves the model to the specified path.
```

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,157 @@
+
+        if isinstance(dataset, DatasetDict):
+            self._train_dataset = dataset["train"]
+            self._eval_dataset = dataset["test"]
```
This might require some documentation somewhere

davidberenstein1957 replied:
How so?

tomaarsen replied:
Not all datasets use `train` and `test` (exclusively). E.g. `dev`, `eval`, `validation`, etc.

davidberenstein1957 replied:
That is true but the dataset is inherited from the `prepare_for_training` argilla method, which only supports "train", "test",  but I will add some comments to clarify.

tomaarsen replied:
Oh, then it should be fine I reckon.

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,157 @@
+        self._label2id = {v: k for k, v in self._id2label.items()}
+
+        self.setfit_model_kwargs = get_default_args(SetFitModel.from_pretrained)
+        self.setfit_model_kwargs["pretrained_model_name_or_path"] = "all-MiniLM-L6-v2"
```
The choice here has notable impact on performance. Do you want people to be able to update this only via `update_config`? It's fine if so, I'm just wondering.

davidberenstein1957 replied:
are you talking about the model? Assuming that we focus on people that want to get an easy approach I wanted to use a quick-and-dirty baseline. 
- What model would you suggest? 
- Perhaps the choice of model can be added as a specific parameter to the `BaseTrainer` to make it more flexible, however, I prefer having most config-related things go via the update config function.

tomaarsen replied:
We tend to use `"sentence-transformers/paraphrase-mpnet-base-v2"` for our examples. It's a bit larger than "all-MiniLM-L6-v2". And I'm fine with just using the `update_config` method to update the params.

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,157 @@
+    ):
+        """
+        It updates the `setfit_model_kwargs` and `setfit_trainer_kwargs` dictionaries with the keyword
+        arguments passed to the `update_config` function
```
```suggestion
        Updates the `setfit_model_kwargs` and `setfit_trainer_kwargs` dictionaries with the keyword
        arguments passed to the `update_config` function.
```
Alternatively, something like "Updates config with keywords passed to `SetFitModel.from_pretrained` and `SetFitTrainer`.", as users don't know what `setfit_model_kwargs` or `setfit_trainer_kwargs` is.

tomaarsen commented on src/argilla/training/utils.py:
```diff
@@ -0,0 +1,56 @@
+#  Copyright 2021-present, the Recognai S.L. team.
```
These are some cool functions. I think I can use something like this for a hyperparameter tuning issue I was facing for SetFit 1.0.

davidberenstein1957 replied:
Nice🤓

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,157 @@
+        self.setfit_trainer_kwargs["model"] = self._model
+        self.__trainer = SetFitTrainer(**self.setfit_trainer_kwargs)
+        self.__trainer.train()
+        self._metrics = self.__trainer.evaluate()
```
I haven't ran any of this yet, but `evaluate` might not work if `eval_dataset` is None.

davidberenstein1957 replied:
That makes sense, I will check.

tomaarsen reviewed (commented):
Looking very solid already. I haven't been able to run this yet (as my PC is mostly running NER experiments 24/7 haha), but I do have a few comments and nitpicks. Also, if this is still WIP, then you can set this PR as a draft, so others know that it's not ready.

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/setfit.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
Update src/argilla/training/base.py

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
fix: resolved comments @tomaarsen

davidberenstein1957 committed:
chore: added additional outline of transformers training module

davidberenstein1957 committed:
chore: initial working version with transformers for multi-label text

davidberenstein1957 committed:
fix: avoid f1 reference overwrite

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,170 @@
+        """
+        return self._trainer.predict(text, as_argilla_records)
+
+    def train(self, path: str = None):
```
Thoughts on `output_dir` instead of `path`? `path` gives me the impression of an "input path" moreso than an output one.
`path` is a fine parameter for `.save()` howver, the method there already implies an output path is meant.

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class ArgillaSetFitTrainer(ArgillaTransformersTrainer):
+    # @require_version("setfit", "0.6")
```
Is this meant to be commented out?

davidberenstein1957 replied:
For now yes, for later, no.

alvarobartt committed:
feat: add `ArgillaSpaCyTrainer` for both `TokenClassification` and `TextClassification` (#2604)

# Description

This PR adds `ArgillaSpaCyTrainer` to use `spaCy` to train both
`token-classification` and `text-classification` models through as of
the recent development of `argilla.training`. `ArgillaSpaCyTrainer`
leverages the features of [`spacy
train`](https://spacy.io/usage/training) programmatically, to simplify
the integration with Argilla.

This PR has been created on top of
https://github.com/argilla-io/argilla/pull/2574 which is the former PR
created by @davidberenstein1957 to add `argilla.training`.

Closes #2594

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

No unit tests have been developed as part of this PR since those will
either be pushed as part of the main PR at
https://github.com/argilla-io/argilla/pull/2574, or on another PR before
merging into `develop`.

I took the chance to create a Jupyter Notebook following the
`template.ipynb` provided by Argilla. The Jupyter Notebook shows how to
train a NER from scratch using `spaCy`, and is available at
https://colab.research.google.com/drive/1zwIpSMMhfyuUmuy2lMpp-HbZc-un8S2N?usp=share_link

**Checklist**

- [X] I have merged the original branch into my forked branch
- [x] I added relevant documentation
- [X] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

davidberenstein1957 committed:
chore: added working version of inference and training for transformers textclassification

alvarobartt commented on src/argilla/training/spacy.py:
```diff
@@ -25,6 +26,8 @@
 
 
 class ArgillaSpaCyTrainer:
+    _logger = logging.getLogger("argilla.training.ArgillaSetFitTrainer")
```
```suggestion
    _logger = logging.getLogger("argilla.training.ArgillaSpaCyTrainer")
```

davidberenstein1957 replied:
I had just updated this too.

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -13,13 +13,16 @@
 
 
 class ArgillaSetFitTrainer(ArgillaTransformersTrainer):
+    _logger = logging.getLogger("argilla.training.ArgillaSetFitTrainer")
```
Is this equivalent to `getLogger(__init__)`? I tend to see that pattern, but this is likely just as fine.

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -100,6 +110,10 @@ def predict(self, text: Union[List[str], str], as_argilla_records: bool = True):
           A list of predictions
         """
+        if self._model is None:
+            self._logger.info("Using model without fine-tuning.")
```
I think `self._logger.warn` is more appropriate here, would you agree?

davidberenstein1957 replied:
Agree

alvarobartt commented on src/argilla/training/spacy.py:
```diff
@@ -177,30 +187,35 @@ def predict(
             str_input = True
 
         formatted_prediction = []
         for doc in self._nlp.pipe(text):
```
Here maybe we could just do something like `raw_preds = list(self._nlp.pipe(text)` and just loop over those if `as_argilla_records=True`, so that we can just skip the manual for-loop when `as_argilla_records=False`

davidberenstein1957 replied:
I agree it is nicer to return the doc generator.

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -112,6 +118,40 @@ def init_args(self):
+
+        from transformers import pipeline
+
+        os.environ["PYTORCH_ENABLE_MPS_FALLBACK"] = "1"
```
Does this allow the CPU to be used automatically if MPS isn't supported for the task? I've never used a Mac, so I don't have experience with MPS myself.

alvarobartt replied:
Yes, indeed that's what the flag means, if the op is not supported for MPS in PyTorch yet, the CPU will be used instead 👍🏻

davidberenstein1957 replied:
Yes, and some architectures don't fully support "mps", so it is a safety measure for that too.

tomaarsen replied:
I see, I see. Hopefully some time in the future, MPS will get the love it deserves.

davidberenstein1957 committed:
fix: resolved comments Tom
chore: added tokenclassification support for transformers

davidberenstein1957 committed:
fix: aligned implementation with comments #2610

davidberenstein1957 committed:
fix: updated formatting

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -258,6 +251,8 @@ def compute_metrics_text_classification(eval_pred):
                 func = compute_metrics_text_classification
 
         elif self._record_class == rg.TokenClassificationRecord:
+            import seqeval as _  # noqa
```
Is this just to trigger an ImportError if it doesn't exist? I think loading `"seqeval"` with evaluate throws an ImportError as well, with some additional context. 
Perhaps we can use `require_dependency` or however that function was called here?

davidberenstein1957 replied:
That was a correct assumption, removed this again.

alvarobartt commented on src/argilla/training/base.py:
```diff
@@ -0,0 +1,216 @@
+                model=self.model,
+                multi_label=self._multi_label,
+                seed=self._seed,
+            )
```
~~Also, this is duplicated 😩 I assume that's my bad due to the merge of my branch with this one.~~

Edit: wrong suggestion!

davidberenstein1957 replied:
Sorry, what is duplicated exactly? @alvarobartt

alvarobartt replied:
Oops my bad, I thought that the `ArgillaSpaCyTrainer` was duplicated but it's not! Feel free to remove this comment!

davidberenstein1957 committed:
fix: removed seqeva import
chore: added initial test outline

davidberenstein1957 committed:
fix: resolved comments Tom and Alvaro

davidberenstein1957 committed:
fix: resolved device misaignment during inference

davidberenstein1957 committed:
chore: initial setup tests with actual training for each framework

davidberenstein1957 committed:
fix: updated conftest formatting

alvarobartt commented on tests/training/conftest.py:
```diff
@@ -13,6 +13,7 @@
 #  limitations under the License.
 
 import argilla
```
Maybe we can just remove this import and replace all the `argilla.` occurrences with `rg.`? WDYT @davidberenstein1957 

```suggestion
```

tomaarsen replied:
Resolved in a324170bb895720290fd2a47f49890524d5bebdd, thanks for the review

alvarobartt commented on tests/training/test_base.py:
```diff
@@ -0,0 +1,69 @@
+
+@pytest.mark.parameterize(["spacy", "transformers", "setfit"])
+def test_base_text_classification_without_split(framework, dataset_text_classification):
+    trainer = ArgillaTrainer(dataset=dataset_text_classification, framework=framework)
```
Maybe we see an underperformance issue with `spaCy` since we're dumping the datasets into a file during the class initialization, which may be cleaner when calling the `train` method instead, as well as the reusability of the `dev` file if `train_size=1`.

Do you agree that I create a quick PR on top of this one to improve those things so that the CI runs faster when integrated? (I think tomorrow afternoon, if not today, the PR can be ready to review)

davidberenstein1957 replied:
I Agree!

alvarobartt replied:
Cool @davidberenstein1957 then I'll try to submit another PR on top of this one later today 😄

alvarobartt replied:
I've created @davidberenstein1957 LMKWYT 🤗 https://github.com/argilla-io/argilla/pull/2642

frascuchon commented on src/argilla/client/datasets.py:
```diff
@@ -718,14 +720,11 @@ def _prepare_for_training_with_transformers(
-        inputs_keys = {key: None for rec in self._records for key in rec.inputs if rec.annotation is not None}.keys()
-
-        ds_dict = {**{key: [] for key in inputs_keys}, "label": []}
+        ds_dict = {"text": [], "label": []}
```
If we keep only the `text` field, we should add some tests with multiple inputs in records for text classification. I can imagine that, for those cases, the `text` field will contain an aggregation of inputs.

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into feat/2504-add-an-argillatraining-module-to-the-core-library-with-training-workflows

tomaarsen committed:
Reword ArgillaTrainer docstring slightly

tomaarsen committed:
Fix parametrize, correct init. in trainer tests

tomaarsen committed:
Remove unnecessary argilla import

alvarobartt committed:
Dump dataset into disk in `train`

alvarobartt committed:
Relocate imports to avoid `require_version` calls when not needed

alvarobartt committed:
Remove `_required_fields` as of recent API changes

tomaarsen committed:
Update phrasing of warning

tomaarsen committed:
Move dataset dump to `train`, ignored unnecessary imports, & remove `_required_fields` attribute (#2642)

# Description

As of recent updates and findings, some things could be improved in
`argilla.training`.

Since the `ArgillaSpaCyTrainer` needs to dump the dataset to disk before
actually training the model, it has been moved out of the `__init__` so
that the dataset/s are dumped once the `train()` method is called.
Additionally, a warning message has been included to let the user know
that the process may take longer due to the dataset dump being a
requirement from `spaCy` training.

Besides that, some other things have been spotted:
* As the `ArgillaSetFitTrainer` and `ArgillaTransformersTrainer` imports
are on the top of the file, those are imported when using any framework
in `ArgillaTrainer`. So on, since some of those require specific
dependencies, we've moved those inside their respective if-statements,
so that e.g. `ArgillaSpaCyTrainer` doesn't require `setfit` to be
installed.
* As of dfa1ee444b42fbe1831f2f3ca111146322385048, the param `fields` has
been removed in `rg.load`, so on, the `_require_fields` attribute set in
`ArgillaTrainer` is no longer valid.

**Type of change**

- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

Since the unit tests are being developed in the target branch, those
have not been created in this PR, but the following script has been
tested successfully!

```python
import argilla as rg
from argilla.training import ArgillaTrainer

rg.init(api_url="https://alvarobartt-argilla-datasets.hf.space", api_key="team.apikey")

trainer = ArgillaTrainer(name="ner", framework="spacy", train_size=0.8)
trainer.update_config(max_epochs=2)
trainer.train(output_dir="ner")
records = trainer.predict("I live in Barcelona.", as_argilla_records=True)

rg.log(
    records=records,
    name="ner",
    tags={
        "task": "NER",
        "family": "token-classification",
        "dataset": "gutenberg-time",
    },
)
```

**Checklist**

- [X] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

tomaarsen committed:
Solve numerous bugs, fix trainer base and SetFit tests

tomaarsen committed:
Improve transformer trainer tests

davidberenstein1957 committed:
fix: added evaluate package to dev tests

davidberenstein1957 committed:
fix: add setft to test dependencies

davidberenstein1957 committed:
fix: added tests for spacy, setfit, transformers

tomaarsen commented on tests/training/helpers.py:
```diff
@@ -0,0 +1,37 @@
+from argilla.training import ArgillaTrainer
+
+
+def cleanup(trainer: ArgillaTrainer, output_dir: str, train: bool = True):
```
I've just found a potential alternative to this:
https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmp-path-factory-fixture

I must say that I couldn't find this earlier when I googled...
This approach is fine too, though.

davidberenstein1957 committed:
fix: corrected test to resolve empty `text` record field with `inputs`

davidberenstein1957 committed:
chore: updated CHANGELOG.md

davidberenstein1957 committed:
chore: added argilla.training docs

tomaarsen commented on docs/_source/guides/train_a_model.ipynb:
It won't let me comment on the exact spot, but the last sentence of `Prepare dataset for training` has a comma where it should not have one.

tomaarsen replied:
Additionally, if you prefer, we can replace the "X" in the table with one of:
* ✅
* ☑️
* ✔️

tomaarsen replied:
Both of these are resolved now, Thanks David

davidberenstein1957 committed:
chore: update docs and process comments @tomaarsen

davidberenstein1957 committed:
fix: updated small typo in docs

davidberenstein1957 committed:
fix: corrected for typo

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "source": [
+    "# 🦾 Train a Model\n",
+    "\n",
+    "This guide showcases how to train a model on the the `Dataset` classes in the Argilla client.\n",
```
```suggestion
    "This guide explains how to train custom models with your annotated data using the `Dataset` classes.\n",
```

nataliaElv replied:
```suggestion
    "This guide showcases how to train a model on the `Dataset` classes in the Argilla client.\n",
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "# 🦾 Train a Model\n",
+    "\n",
+    "This guide showcases how to train a model on the the `Dataset` classes in the Argilla client.\n",
+    "The Dataset classes are lightweight containers for Argilla records. These classes facilitate importing from and exporting to different formats (e.g., `pandas.DataFrame`, `datasets.Dataset`) as well as sharing and versioning Argilla datasets using the Hugging Face Hub. For more info, take a look at the\n",
```
Something is missing in the last sentence

nataliaElv replied:
Is there a missing link at the end of this line?

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "id": "ae9d4c9e-24a1-4a59-8a17-3b6ac1a39c88",
+   "metadata": {},
+   "source": [
+    "## Train custom workflow \n",
```
```suggestion
    "## Train with a custom workflow \n",
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "## Train custom workflow \n",
+    "\n",
+    "### Prepare for training\n",
+    "If you want to train a model we provide a handy method to prepare your dataset: `DatasetFor*.prepare_for_training()`.\n",
```
```suggestion
    "If you want to train a model, Argilla provides a handy method to prepare your dataset: `DatasetFor*.prepare_for_training()`.\n",
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "\n",
+    "### Prepare for training\n",
+    "If you want to train a model we provide a handy method to prepare your dataset: `DatasetFor*.prepare_for_training()`.\n",
+    "It will return a Hugging Face dataset, a spaCy DocBin or a SparkNLP-formatted DataFrame, optimized for the training process with the Hugging Face Trainer, the spaCy CLI or the SparkNLP API. Our [libraries deepdive](../libraries/libraries.html) and [training tutorials](../../tutorials/steps/2_training.md) show entire training workflows for your favorite packages."
```
```suggestion
    "It will return a Hugging Face dataset, a spaCy DocBin, or a SparkNLP-formatted DataFrame, optimized for the training process with the Hugging Face Trainer, the spaCy CLI, or the SparkNLP API respectively. The [libraries deepdive](../libraries/libraries.html) and [training tutorials](../../tutorials/steps/2_training.md) show entire training workflows for your favorite packages."
```

nataliaElv replied:
These two links aren't working for me.

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "source": [
+    "### Train-test split\n",
+    "\n",
+    "It is possible to directly include train-test splits to the `prepare_for_training` by passing the `train_size` and `test_size` parameters."
```
```suggestion
    "It's possible to create train-test splits to the `prepare_for_training` by passing the `train_size` and `test_size` parameters."
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "\n",
+    "*TextClassification*\n",
+    "\n",
+    "For text classification tasks, it flattens the inputs into separate columns of the returned dataset and converts the annotations of your records into integers and writes them in a label column:\n",
```
```suggestion
    "For text classification tasks, it flattens the inputs into separate columns of the returned dataset, converts the annotations of your records into integers and writes them in a label column:\n",
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "source": [
+    "### The `ArgillaTrainer`\n",
+    "\n",
+    "We can use the `ArgillaTrainer` to train directly for frameworks by `spacy`, `setfit` and `transformers` as framework variables."
```
```suggestion
    "You can use the `ArgillaTrainer` to train directly for frameworks by `spacy`, `setfit` and `transformers` as framework variables."
```

nataliaElv replied:
```suggestion
    "We can use the `ArgillaTrainer` to train directly using `spacy`, `setfit` and `transformers` as framework variables."
```

nataliaElv commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "source": [
+    "## Train directly\n",
+    "\n",
+    "The `ArgillaTrainer` is an wrapper around many of our favorite NLP libraries, ad provides a very intuitive abstract workflow to facilitate simple training workflows using decent default pre-set configuration without having to worry about any data transformations from Argilla. We plan on adding more support for tasks and frameworks like OpenAI, AutoTrain and SageMaker during coming releases.\ ...
```
```suggestion
    "The `ArgillaTrainer` is a wrapper around many of our favorite NLP libraries. It provides a very intuitive abstract workflow to facilitate simple training workflows using decent default pre-set configurations without having to worry about any data transformations from Argilla. We plan on adding more support for tasks and frameworks like OpenAI, AutoTrain and SageMaker in the coming releases.\n",
```

dvsrepo commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "\n",
+    "For each record type there's a corresponding Dataset class called `DatasetFor<RecordType>`.\n",
+    "You can look up their API in the [reference section](../reference/python/python_client.rst#module-argilla.client.datasets)."
+   ]
```
I would add a brief outline of the two main sections. 

Something like:

There are two ways to train custom models on top of your annotated data:

1. Train models using the Argilla training module.
2. Train with a custom workflow using the prepare for training methods

I would change the order of the sections, i.e., Argilla trainer should go first. Also if you feel like adding a sentence about when to use what in the outline that's great. Something like: Custom workflows give you more flexibility to integrate with your existing training workflows.

dvsrepo reviewed (changes requested):
This is looking good! I've added some suggestions.

One last thing I'm missing is to configure the Python reference index to add the Training module as we have for the other modules, so the docstring docs are browsable there

nataliaElv commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "source": [
+    "### Update training config\n",
+    "\n",
+    "The trainer also has an `ArgillaTrainer.update_config()` method, which maps `**kwrags` to the respective framework. So, these can be derived from the underlying framework that was used to initialize the trainer. Underneath, an overview of these variables is given for the supported frameworks. Note that, you don't need to pass all of them directly and that the values below are their default c ...
```
```suggestion
    "The trainer also has an `ArgillaTrainer.update_config()` method, which maps `**kwargs` to the respective framework. So, these can be derived from the underlying framework that was used to initialize the trainer. Underneath, you can find an overview of these variables for the supported frameworks. Note that you don't need to pass all of them directly and that the values below are their default configurations."
```

nataliaElv commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+    "    seed = 42,\n",
+    "    use_amp = True,\n",
+    "    warmup_proportion = 0.1,\n",
+    "    distance_metric = `BatchHardTripletLossDistanceFunction.cosine_distance`,\n",
```
Are `` the correct quotemarks here?

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

davidberenstein1957 committed:
Update docs/_source/guides/train_a_model.ipynb

Co-authored-by: Daniel Vila Suero <daniel@recogn.ai>

nataliaElv commented on docs/_source/guides/train_a_model.ipynb:
```diff
@@ -0,0 +1,358 @@
+   "id": "9cfa853d",
+   "metadata": {},
+   "source": [
+    "#### SetFit\n",
```
For some reason I can't see this snippet with Python format in the docs. But the others do work.

nataliaElv reviewed (commented):
Just some comments and suggestions for the doc update

davidberenstein1957 committed:
fix: resolved comments dani

frascuchon committed:
Avoid changing environment deps

tomaarsen commented on src/argilla/training/setfit.py:
```diff
@@ -0,0 +1,161 @@
+        if self._eval_dataset:
+            self._metrics = self.__trainer.evaluate()
+        else:
+            self._metrics = None
```
These are never used, is that right? Same for transformers.
This is what I get when training, by the way:
<details><summary>Training logs</summary>

```
[04/05/23 12:12:35] WARNING  WARNING:ArgillaTrainer:ArgillaBaseTrainer info:                                                                                               base.py:140
                             _________________________________________________________________
                             These baseline params are fixed:
                                 dataset: research_titles_multi_label
                                 task: DatasetForTextClassification
                                 multi_label: True
                                 train_size: 0.8
                                 seed: None

                             <class 'argilla.training.setfit.ArgillaSetFitTrainer'> info:
                             _________________________________________________________________
                             The parameters are configurable via `trainer.update_config()`:
                                 'SetFitModel'
                             force_download: False
                             resume_download: False
                             proxies: None
                             token: None
                             cache_dir: None
                             local_files_only: False
                             has_token: None
                             arg_name: None
                             arg_value: None
                             pretrained_model_name_or_path: all-MiniLM-L6-v2
                             multi_target_strategy: one-vs-rest
                             device: cpu
                             'SetFitTrainer'
                             model: None
                             train_dataset: Dataset({
                                 features: ['id', 'text', 'label', 'binarized_label'],
                                 num_rows: 14
                             })
                             eval_dataset: Dataset({
                                 features: ['id', 'text', 'label', 'binarized_label'],
                                 num_rows: 4
                             })
                             model_init: None
                             metric: accuracy
                             loss_class: <class 'sentence_transformers.losses.CosineSimilarityLoss.CosineSimilarityLoss'>
                             num_iterations: 20
                             num_epochs: 1
                             learning_rate: 2e-05
                             batch_size: 16
                             seed: 42
                             column_mapping: {'text': 'text', 'binarized_label': 'label'}
                             use_amp: False
                             warmup_proportion: 0.1
                             distance_metric: <function BatchHardTripletLossDistanceFunction.cosine_distance at 0x0000016D3312C048>
                             margin: 0.25
                             samples_per_label: 2

                             Using the trainer:
                             _________________________________________________________________
                             `trainer.train(output_dir)` to train to start training. `output_dir` is the directory to save the model automatically.
                             `trainer.predict(text, as_argilla_records=True)` to make predictions.
                             `trainer.save(output_dir)` to save the model manually.
config.json not found in HuggingFace Hub
                    WARNING  WARNING:huggingface_hub.hub_mixin:config.json not found in HuggingFace Hub                                                               hub_mixin.py:184
model_head.pkl not found on HuggingFace Hub, initialising classification head with random weights. You should TRAIN this model on a downstream task to use it for predictions and inference.
Applying column mapping to training dataset
***** Running training *****
  Num examples = 300
  Num epochs = 1
  Total optimization steps = 19
  Total train batch size = 16
Iteration: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 19/19 [00:06<00:00,  2.78it/s] 
Epoch: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:06<00:00,  6.85s/it] 
Applying column mapping to evaluation dataset
***** Running evaluation *****
```
</details>
The evaluation output is never printed or displayed.

davidberenstein1957 replied:
I fixed this.

davidberenstein1957 committed:
fix: added INFO logging and metric overview

frascuchon commented on CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `argilla.tasks.database.migrate` new task to execute database migrations ([#2564]).
 - `release.Dockerfile` and `quickstart.Dockerfile` now creates a default `argilladata` volume to persist data ([#2564]).
 - Add user settings page. Closes [#2496](https://github.com/argilla-io/argilla/issues/2496)
+- Added `Argilla.training` module with support for `spacy`, `setfit`, and `transformers`. Closes [#2504](https://github.com/argilla-io/argilla/issues/2496)
```
```suggestion
- Added `Argilla.training` module with support for `spacy`, `setfit`, and `transformers`. Closes [#2504](https://github.com/argilla-io/argilla/issues/2496)

### Fixes
- Now the `prepare_for_training` method is working when `multi_label=True`. Closes [#2606](https://github.com/argilla-io/argilla/issues/2606)
```

davidberenstein1957 committed:
Update CHANGELOG.md

Co-authored-by: Francisco Aranda <francis@argilla.io>
